### ISSUE TITLE
midtraining learning rate is backwards (bad bug)

### ISSUE THREAD

**Issue by karpathy** [karpathy/nanochat#68]
Slightly embarrassing I just noticed that the learning rate decay is backwards in the midtraining script.
The variable `progress` increases from 0 to 1, while the learning rate multiplier should do the opposite. It should basically be `1 - progress`, because the `lrm` multiplies the init learning rate. Both SFT and RL scripts do it correctly.

https://github.com/karpathy/nanochat/blob/67aaca98f53a6db9f1f5a2a62893f98d443a090a/scripts/mid_train.py#L143-L144

TLDR instead of ramping down the LR I am ramping it up. Classic example in the wild of [neural networks fail silently](https://karpathy.github.io/2019/04/25/recipe/). For the record I would like to blame Cursor autocomplete for this bug :).

This probably also explains why there is such a big gap between midtraining and SFT, which I was a little bit confused by earlier. I basically expected them to be very close to each other and I wasn't even sure why SFT is necessary exactly. Because SFT ramps down correctly, we end up with an effectively triangular learning rate overall, where we first ramp up (midtraining) and then ramp down (sft). This is not too uncommon to do intentionally (having a warmup period), but we certainly should do it intentionally rather than unintentionally.

I'm doing some simple sweeps now to re-tune the midtraining hyperparameters for the speedrun.

Creating issue so that I can reference it in the fix commit.

**Comment by karpathy** [comment]
Update it looks like:

<img width="1168" height="326" alt="Image" src="https://github.com/user-attachments/assets/aa403dfa-3b9f-4afd-8db6-748852c76c9a" />

1. ramping the LR up (current, gray) is clearly not the best idea lol
2. simply keeping LR multiplier at 1 (pretraining LR) then decaying for last 20% seems to work best, outperforming linear rampdown and rampdown over last 10%.
3. brief warmup period seems to not be helping

Merging the 20% version shortly.
